faunadb-java-dsl/src/main/java/com/faunadb/client/types/Types.java

```diff
@@ -0,0 +1,231 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.errors.FaunaException;
+import org.joda.time.Instant;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.lang.String.format;
+
+/**
+ * Auxiliary class used to create types representation while decoding user defined objects.
+ */
+public final class Types {
```

Comment:
This has nothing to do with codecs, this is to overcome a limitation of jvm with type erasure, since `Map`s and `Collection`s classes don't carry the information of the underlying types I made a subclass of `java.lang.reflect.Type` that carry that information (`CollectionType` and `MapType`). `SimpleType` is a tag type to identify and stop introspection on those types. This is how decoder identify the correct types for maps/collections. This is a simplified version of Jackson. Adding that feature on `Value` would be just a sugar on top of `Decoder`